### ISSUE TITLE
Discussion: Alternative custom primitive implementation

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#2482]
I would like to discuss an alternative approach of our DistributedLog primitive.
Please share your thoughts about this.

### Idea

 We have our custom primitive, but instead of writing the events to a file we directly process it.
 
 #### Processing
 
 We start the processing in the primitive, this means we run the processor on leader and follower, so we process and build state on all nodes. The snapshot creation is triggered by the primitive. On large snapshot sizes it makes sense only to write in the atomix snapshot the table of contents and write the snapshot separate.
 On snapshot replication (triggered by atomix), for small state the rocksdb content is in the atomix snapshot. If the state is larger then 1G, the atomix snapshot contains only a table of contents, which is send to the follower. 
 The follower then needs to request the snapshot from the leader (we send it via atomix).

Since the follower also process the events we need to handle the case where it tries to append new entries. Here we have different solutions:

 1. Use strategy pattern and the raft listeners:
    * This means we listen for the raft events and only on the leader we install the real DistributionLogClient. On the follower a NoopLogClient is installed, which does nothing.
 2. Only append Commands to the log:
    * We only append Commands by the ZeebeClient to the atomix log, which are then processed. 
     Events are then only given to the exporters, which should be idempotent anyway.
 3. Write leader and follower events:
    * On processing we then only process leader events
 
 #### Re-processing
 
 Primitive restarts from snapshot and events are replayed by atomix. Our state contains last processed position, so no events/commands are processed twice.
 
 If follower was long gone and comes back, snapshot and log is replicate by atomix. Primitive should block until our own snapshot replication is done. After the replication, the log is replayed on the follower and position from state is used to not process events twice.
 
 #### Exporting
 
 The primitive starts on creation an separate thread with an one to one queue. These queue is filled by the primitive with
 new events. The thread orchestrates the exporters and consumes the queue. The exporter threads runs on all nodes.
 This means we need to make sure that the exporters are **idempotent**, but this gives us the following:
 
  * all nodes will process and export the records
  * if one node dies: records are still processed and exported by other nodes
  * if quorum dies: remaining records, which are already committed, are processed and exported - no new events can be committed
  
  #### Benefites
  
  * no own logstream management
    * no need for own file backed data structures 
    * no need to delete data by our own
    * no need to replicate the log by our own
  * distinction between leader and followers only on appending
     * if we choose the second solution, we have no distinction at all

  * Performance ?
   * Depending on the approach we choose for the appending we can gain or lose performance
     * If we choose the first approach there should be no difference to our old own RAFT and Processing implementation (except that we probably have more events now), since new commands can be still appended to the log. But these events/commands are one by one applied to the state machine / custom primitive (processed by us).
     * If we choose the second approach we can gain a lot of performance, since we have no events to write and commit, they are directly processed and exported
     * If we choose the third approach we lose a lot of performance, since the event count is multiplied by the count of RAFT members
   * I assume we gain better performance compared to our current approach since one layer (appending to own log) disappears
   * But this has to be measured 
  
#### Requires

 * state replication needs to be done by our self on large scenarios (we have this functionallity already)
 * exporters needs to be idempotent (this is already requested regarding the docs)

**Comment by npepinpe** [comment]
:+1: 

In general, I like the approach, but I'm still not 100% sure that we can maintain the same at-least-once guarantee with exporters: how do we replay events that haven't been exported because exporters were slow? Let's say my exporter is super slow, and nodes fail in a round-robin fashion: wouldn't I lose events there? Unless we don't mark a record as processed until it has been exported, but then how would an exporter implement batching/some strategy to not DDoS the endpoint it is exporting to?

If we use the Atomix log only, I think we cannot use this last-exported-position to control where the replay start; and if we could, we would still need to make sure Atomix doesn't truncate the log (afaik it just uses the last processed command position).

But as I said I like it, and I wish we didn't have to build all this replication stuff around Atomix. I think it's worth discussing alternatives, and thanks for spending time doing that on a Sunday ;)

Whatever we decide, I think our current approach is definitely less than ideal so we should find an alternative.

**Comment by npepinpe** [comment]
One thing to note regarding the exporter issue, @deepthidevaki and I would like to see how much effort it would be to control at which position Atomix can truncate its log/should start replaying, so that would maybe solve the issue.

**Comment by Zelldon** [comment]
Thanks for sharing your thoughts. I'm glad that I'm not the only one which is not 100% satisfied with the current approach.

I think with your comment you hit the nerve. This was also mentioned by @menski that we could have problems if the exporter lays behind. But maybe we can find a solution for this as well.

I have two ideas to solve this:

1. We do it the same way as we do it currently, but only for the exporters. This means we write a log, which is consumed by the exporters and where we can delete data after the last exporter has consumed it. But still we also have to do the replication, which feels weird that we build a part of RAFT around a RAFT implementation. Might make everything a bit easier because we only have one consumer, the exporter, not both.

2. We use a separate custom primitive for this. This was part of the first idea above, but I discarded this, because I thought a separate thread for exporting could be enough. I think it is not easy to create/get a primitive inside a primitive, but i think we can solve this via dependency injection.  In our first primitive we give the record to our Processor and append it also to the exporter primitive or we pass events to the exporter primitive during the processing. The exporter primitive then only exports the records one by one. Atomix will then do the replication and deletion part.

**Comment by npepinpe** [comment]
I thought about it a little bit. 

Why not implement it the same way we did our exporters? This means we need to:

1. Provide access to an Atomix `RaftLogReader`, which is do-able by obtaining the `RaftPartitionGroup` (after Atomix has started), then getting the partition; we would need to add two levels of delegation, such that the `RaftPartition#openReader(index)` calls its `RaftPartitionServer#openReader(index)` which would call the currently existing `RaftServer#openReader(index)`, and while I don't think those changes would necessarily be merged upstream, they're very small, and limited to delegating to pre-existing members.
1. I already poc'd a way we could control the raft compaction; there is a single point where when a `RaftServer` takes a snapshot, it calls `raft.getLastAppliedIndex()` (https://github.com/zeebe-io/atomix/blob/0ac81aeb3d13603434462e0e454fb054accbf912/protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftServiceManager.java#L171). We introduce an object which controls the compaction index; it is notified when a new entry is applied, and it is notified when one is exported. That way when snapshotting we can provide the correct index to compact at.

I did a quick and dirty poc in a branch [here](https://github.com/zeebe-io/atomix/tree/raft-controlled-compaction/tests/src/main/java/io/atomix/protocols/raft/test/exportable). This doesn't exactly work 100% but it shows the basic functionality and I think it would be possible to get rid of our log and restore stuff. We'd probably still need to do snapshotting on top (though the latest PR removes the maximum size of a snapshot so we can revisit that if we think simpler is worth the disk space).
